The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10067672375/job/27831612376) [(plain)](https://github.com/rust-lang/rust/commit/036b4f075a062bda29bf37137f0110a236945fa1/checks/27831612376/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.679 Building wheels for collected packages: reuse
#16 2.680   Building wheel for reuse (pyproject.toml): started
#16 2.928   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 2.929   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 2.929   Stored in directory: /tmp/pip-ephem-wheel-cache-klh3x_4d/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 2.932 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.325 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.325 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.4s
---
    Checking rustc-demangle v0.1.24
error[E0658]: use of unstable library feature 'async_closure'
   --> library/alloc/src/boxed.rs:202:17
    |
202 | use core::ops::{AsyncFn, AsyncFnMut, AsyncFnOnce};
    |
    = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
    = help: add `#![feature(async_closure)]` to the crate attributes to enable
    = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date
    = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'async_closure'
   --> library/alloc/src/boxed.rs:202:26
    |
202 | use core::ops::{AsyncFn, AsyncFnMut, AsyncFnOnce};
    |
    = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
    = help: add `#![feature(async_closure)]` to the crate attributes to enable
    = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date
    = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'async_closure'
   --> library/alloc/src/boxed.rs:202:38
    |
202 | use core::ops::{AsyncFn, AsyncFnMut, AsyncFnOnce};
    |
    = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
    = help: add `#![feature(async_closure)]` to the crate attributes to enable
    = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date
    = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'async_closure'
    --> library/alloc/src/boxed.rs:2168:22
     |
2168 | impl<Args: Tuple, F: AsyncFnOnce<Args> + ?Sized, A: Allocator> AsyncFnOnce<Args> for Box<F, A> {
     |
     = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
     = help: add `#![feature(async_closure)]` to the crate attributes to enable
     = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date
     = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'async_closure'
    --> library/alloc/src/boxed.rs:2168:64
     |
2168 | impl<Args: Tuple, F: AsyncFnOnce<Args> + ?Sized, A: Allocator> AsyncFnOnce<Args> for Box<F, A> {
     |
     = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
     = help: add `#![feature(async_closure)]` to the crate attributes to enable
     = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date
     = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'async_closure'
    --> library/alloc/src/boxed.rs:2178:22
     |
2178 | impl<Args: Tuple, F: AsyncFnMut<Args> + ?Sized, A: Allocator> AsyncFnMut<Args> for Box<F, A> {
     |
     = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
     = help: add `#![feature(async_closure)]` to the crate attributes to enable
     = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date
     = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'async_closure'
    --> library/alloc/src/boxed.rs:2178:63
     |
2178 | impl<Args: Tuple, F: AsyncFnMut<Args> + ?Sized, A: Allocator> AsyncFnMut<Args> for Box<F, A> {
     |
     = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
     = help: add `#![feature(async_closure)]` to the crate attributes to enable
     = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date
     = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'async_closure'
    --> library/alloc/src/boxed.rs:2187:22
     |
2187 | impl<Args: Tuple, F: AsyncFn<Args> + ?Sized, A: Allocator> AsyncFn<Args> for Box<F, A> {
     |
     = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
     = help: add `#![feature(async_closure)]` to the crate attributes to enable
     = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date
     = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'async_closure'
    --> library/alloc/src/boxed.rs:2187:60
     |
2187 | impl<Args: Tuple, F: AsyncFn<Args> + ?Sized, A: Allocator> AsyncFn<Args> for Box<F, A> {
     |
     = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
     = help: add `#![feature(async_closure)]` to the crate attributes to enable
     = note: this compiler was built on 2024-06-10; consider upgrading it if it is out of date
```

</details>

--- a/compiler/rustc_ast_lowering/src/path.rs
+++ b/compiler/rustc_ast_lowering/src/path.rs
@@ -44,13 +44,11 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         let mut res = self.lower_res(base_res);
 
         // When we have an `async` kw on a bound, map the trait it resolves to.
-        let mut bound_modifier_allowed_features = None;
         if let Some(TraitBoundModifiers { asyncness: BoundAsyncness::Async(_), .. }) = modifiers {
             match res {
                 Res::Def(DefKind::Trait, def_id) => {
-                    if let Some((async_def_id, features)) = self.map_trait_to_async_trait(def_id) {
+                    if let Some(async_def_id) = self.map_trait_to_async_trait(def_id) {
                         res = Res::Def(DefKind::Trait, async_def_id);
-                        bound_modifier_allowed_features = Some(features);
                     } else {
                         self.dcx().emit_err(AsyncBoundOnlyForFnTraits { span: p.span });
                     }
@@ -66,6 +64,16 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                 }
             }
         }
+
+        // Ungate the `async_fn_traits` feature in the path if the trait is
+        // named via either `async Fn*()` or `AsyncFn*()`.
+        let bound_modifier_allowed_features = if let Res::Def(DefKind::Trait, async_def_id) = res
+            && self.tcx.async_fn_trait_kind_from_def_id(async_def_id).is_some()
+        {
+            Some(self.allow_async_fn_traits.clone())
+        } else {
+            None
+        };
 
         let path_span_lo = p.span.shrink_to_lo();
         let proj_start = p.segments.len() - unresolved_segments;
@@ -506,14 +514,14 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
     /// This only needs to be done until we unify `AsyncFn` and `Fn` traits into one
     /// that is generic over `async`ness, if that's ever possible, or modify the
     /// lowering of `async Fn()` bounds to desugar to another trait like `LendingFn`.
-    fn map_trait_to_async_trait(&self, def_id: DefId) -> Option<(DefId, Lrc<[Symbol]>)> {
+    fn map_trait_to_async_trait(&self, def_id: DefId) -> Option<DefId> {
         let lang_items = self.tcx.lang_items();
         if Some(def_id) == lang_items.fn_trait() {
-            Some((lang_items.async_fn_trait()?, self.allow_async_fn_traits.clone()))
+            lang_items.async_fn_trait()
         } else if Some(def_id) == lang_items.fn_mut_trait() {
-            Some((lang_items.async_fn_mut_trait()?, self.allow_async_fn_traits.clone()))
+            lang_items.async_fn_mut_trait()
         } else if Some(def_id) == lang_items.fn_once_trait() {
-            Some((lang_items.async_fn_once_trait()?, self.allow_async_fn_traits.clone()))
+            lang_items.async_fn_once_trait()
         } else {
             None
         }

--- a/library/alloc/src/lib.rs
+++ b/library/alloc/src/lib.rs
@@ -101,6 +101,7 @@
 #![feature(array_windows)]
 #![feature(ascii_char)]
 #![feature(assert_matches)]
+#![feature(async_closure)]
 #![feature(async_fn_traits)]
 #![feature(async_iterator)]
 #![feature(clone_to_uninit)]

--- a/library/core/src/ops/async_function.rs
+++ b/library/core/src/ops/async_function.rs
@@ -4,7 +4,7 @@ use crate::marker::Tuple;
 /// An async-aware version of the [`Fn`](crate::ops::Fn) trait.
 ///
 /// All `async fn` and functions returning futures implement this trait.
-#[unstable(feature = "async_fn_traits", issue = "none")]
+#[unstable(feature = "async_closure", issue = "62290")]
 #[rustc_paren_sugar]
 #[fundamental]
 #[must_use = "async closures are lazy and do nothing unless called"]
@@ -18,7 +18,7 @@ pub trait AsyncFn<Args: Tuple>: AsyncFnMut<Args> {
 /// An async-aware version of the [`FnMut`](crate::ops::FnMut) trait.
 ///
 /// All `async fn` and functions returning futures implement this trait.
-#[unstable(feature = "async_fn_traits", issue = "none")]
+#[unstable(feature = "async_closure", issue = "62290")]
 #[rustc_paren_sugar]
 #[fundamental]
 #[must_use = "async closures are lazy and do nothing unless called"]
@@ -39,7 +39,7 @@ pub trait AsyncFnMut<Args: Tuple>: AsyncFnOnce<Args> {
 /// An async-aware version of the [`FnOnce`](crate::ops::FnOnce) trait.
 ///
 /// All `async fn` and functions returning futures implement this trait.
-#[unstable(feature = "async_fn_traits", issue = "none")]
+#[unstable(feature = "async_closure", issue = "62290")]
 #[rustc_paren_sugar]
 #[fundamental]
 #[must_use = "async closures are lazy and do nothing unless called"]

--- a/tests/ui/async-await/async-fn/edition-2015.rs
+++ b/tests/ui/async-await/async-fn/edition-2015.rs
@@ -3,5 +3,7 @@ fn foo(x: impl async Fn()) -> impl async Fn() { x }
 //~| ERROR `async` trait bounds are only allowed in Rust 2018 or later
 //~| ERROR async closures are unstable
 //~| ERROR async closures are unstable
+//~| ERROR use of unstable library feature 'async_closure'
+//~| ERROR use of unstable library feature 'async_closure'
 
 fn main() {}

--- a/tests/ui/async-await/async-fn/edition-2015.stderr
+++ b/tests/ui/async-await/async-fn/edition-2015.stderr
@@ -38,6 +38,26 @@ LL | fn foo(x: impl async Fn()) -> impl async Fn() { x }
    = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
    = help: to use an async block, remove the `||`: `async {`
 
-error: aborting due to 4 previous errors
+error[E0658]: use of unstable library feature 'async_closure'
+  --> $DIR/edition-2015.rs:1:22
+   |
+LL | fn foo(x: impl async Fn()) -> impl async Fn() { x }
+   |                      ^^^^
+   |
+   = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
+   = help: add `#![feature(async_closure)]` to the crate attributes to enable
+   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
+
+error[E0658]: use of unstable library feature 'async_closure'
+  --> $DIR/edition-2015.rs:1:42
+   |
+LL | fn foo(x: impl async Fn()) -> impl async Fn() { x }
+   |                                          ^^^^
+   |
+   = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
+   = help: add `#![feature(async_closure)]` to the crate attributes to enable
+   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
+
+error: aborting due to 6 previous errors
 
 For more information about this error, try `rustc --explain E0658`.

--- a/tests/ui/async-await/async-fn/simple.rs
+++ b/tests/ui/async-await/async-fn/simple.rs
@@ -2,7 +2,7 @@
 //@ edition: 2021
 //@ build-pass
 
-#![feature(async_fn_traits)]
+#![feature(async_closure)]
 
 extern crate block_on;
 

